process/block/preprocess/smartContractResults.go

```diff
@@ -0,0 +1,424 @@
+	hasher hashing.Hasher,
+	marshalizer marshal.Marshalizer,
+	scrProcessor process.SmartContractResultProcessor,
+	shardCoordinator sharding.Coordinator,
+	accounts state.AccountsAdapter,
+	onRequestSmartContractResult func(shardID uint32, txHashes [][]byte),
+) (*smartContractResults, error) {
+
+	if hasher == nil {
+		return nil, process.ErrNilHasher
+	}
+	if marshalizer == nil {
+		return nil, process.ErrNilMarshalizer
+	}
+	if scrDataPool == nil {
+		return nil, process.ErrNilScrDataPool
+	}
+	if store == nil {
+		return nil, process.ErrNilScrStorage
+	}
+	if scrProcessor == nil {
+		return nil, process.ErrNilTxProcessor
+	}
+	if shardCoordinator == nil {
+		return nil, process.ErrNilShardCoordinator
+	}
+	if accounts == nil {
+		return nil, process.ErrNilAccountsAdapter
+	}
+	if onRequestSmartContractResult == nil {
+		return nil, process.ErrNilRequestHandler
+	}
+
+	bpp := &basePreProcess{
+		hasher:           hasher,
+		marshalizer:      marshalizer,
+		shardCoordinator: shardCoordinator,
+	}
+
+	scr := &smartContractResults{
+		basePreProcess:               bpp,
+		storage:                      store,
+		scrPool:                      scrDataPool,
+		onRequestSmartContractResult: onRequestSmartContractResult,
+		scrProcessor:                 scrProcessor,
+		accounts:                     accounts,
+	}
+
+	scr.chRcvAllScrs = make(chan bool)
+	scr.scrPool.RegisterHandler(scr.receivedSmartContractResult)
+	scr.scrForBlock.txHashAndInfo = make(map[string]*txInfo)
+
+	return scr, nil
+}
+
+// waitForScrHashes waits for a call whether all the requested smartContractResults appeared
+func (scr *smartContractResults) waitForScrHashes(waitTime time.Duration) error {
+	select {
+	case <-scr.chRcvAllScrs:
+		return nil
+	case <-time.After(waitTime):
+		return process.ErrTimeIsOut
+	}
+}
+
+// IsDataPrepared returns non error if all the requested smartContractResults arrived and were saved into the pool
+func (scr *smartContractResults) IsDataPrepared(requestedScrs int, haveTime func() time.Duration) error {
+	if requestedScrs > 0 {
+		log.Info(fmt.Sprintf("requested %d missing scr\n", requestedScrs))
+		err := scr.waitForScrHashes(haveTime())
+		scr.scrForBlock.mutTxsForBlock.RLock()
+		missingScrs := scr.scrForBlock.missingTxs
+		scr.scrForBlock.mutTxsForBlock.RUnlock()
+		log.Info(fmt.Sprintf("received %d missing scr\n", requestedScrs-missingScrs))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RemoveTxBlockFromPools removes smartContractResults and miniblocks from associated pools
+func (scr *smartContractResults) RemoveTxBlockFromPools(body block.Body, miniBlockPool storage.Cacher) error {
+	if body == nil {
+		return process.ErrNilTxBlockBody
+	}
+
+	err := scr.removeDataFromPools(body, miniBlockPool, scr.scrPool, block.SmartContractResultBlock)
+
+	return err
+}
+
+// RestoreTxBlockIntoPools restores the smartContractResults and miniblocks to associated pools
+func (scr *smartContractResults) RestoreTxBlockIntoPools(
```

Comment:
This method is not tested at all

Reply:
were copied from blockprocessor, not tested there with unit tests. Unit tests will be added. Added TODO

Reply:
ok